tests/integration_tests/functional/test_logging.py

```diff
@@ -6,69 +6,88 @@
-    """Convert a pretty-print log level into the related log level code.
-
-    Turns a pretty formatted log level (i.e Warning) into the one actually
-    being logged (i.e WARN).
-    :param log_level: pretty formatted log level
-    :return: actual level being logged
-    """
-    if log_level == "Error":
-        return LOG_LEVELS[0]
-    if log_level == "Warning":
-        return LOG_LEVELS[1]
-    if log_level == "Info":
-        return LOG_LEVELS[2]
-    if log_level == "Debug":
-        return LOG_LEVELS[3]
-    return ""
+    if level == "ERROR":
+        return 0
+    if level == "WARN":
+        return 1
+    if level == "INFO":
+        return 2
+    if level == "DEBUG":
+        return 3
+    if level == "TRACE":
+        return 4
+    raise Exception("Invalid level")
 
 
-def check_log_message_format(log_str, instance_id, level, show_level, show_origin):
+# pylint: disable=anomalous-backslash-in-string
+def check_log_message_format(log_str, instance_id, log_level, show_level, show_origin):
     """Ensure correctness of the logged message.
 
     Parse the string representing the logs and look for the parts
     that should be there.
-    The log line should look lie this:
-         YYYY-MM-DDTHH:MM:SS.NNNNNNNNN [ID:THREAD:LEVEL:FILE:LINE] MESSAGE
-    where LEVEL and FILE:LINE are both optional.
-    e.g. with THREAD NAME as TN
-    `2018-09-09T12:52:00.123456789 [MYID:TN:WARN:/path/to/file.rs:52] warning`
-    """
-    timestamp, tag_and_msg = log_str.split(" ", maxsplit=1)
-    timestamp = timestamp[:-10]
-    strptime(timestamp, "%Y-%m-%dT%H:%M:%S")
-
-    pattern = "\\[(" + instance_id + ")"
-    pattern += ":(.*)"
-    if show_level:
-        pattern += ":(" + "|".join(LOG_LEVELS) + ")"
-    if show_origin:
-        pattern += ":([^:]+/[^:]+):([0-9]+)"
-    pattern += "\\].*"
+    The log line should look like:
+    > {year}-{month}-{day}T{hour}:{minute}:{second}.{nanosecond} [{instance id}:{thread name}:{level}:{file}:{line number}] {message}
+    where `level`, `file` and `line number` are optional e.g.
+    > 2023-07-19T12:10:54.123608814 [fc_api:INFO:src\main.rs:18] yak shaving completed.
+    """
+    split = iter(log_str.split())
+    now = datetime.datetime.now()
+
+    # 2023-07-19T12:10:54.123608814
+    timestamp = next(split)
+    date, time = timestamp.split("T")
+    year, month, day = date.split("-")
+    assert len(month) == 2
+    assert len(day) == 2
+
+    hour, minute, secs = time.split(":")
+    second, nanosecond = secs.split(".")
+    assert len(hour) == 2
+    assert len(minute) == 2
+    assert len(second) == 2
+    assert len(nanosecond) == 9
+
+    # Assert the time in the logs is less than or equal to the current time
+    log_time = datetime.datetime(
+        year=int(year),
+        month=int(month),
+        day=int(day),
+        hour=int(hour),
+        minute=int(minute),
+        second=int(second),
+        microsecond=int(nanosecond) // 1000,
+    )
+    assert log_time <= now
```

Comment:
Why not keep using strptime? If you want to fix the time resolution, you can do it in a separate commit, like this:

```python
timestamp = "2023-12-21T13:12:23.123456789"
strptime(timestamp[:-3], "%Y-%m-%dT%H:%M:%S.%f")
time.struct_time(tm_year=2023, tm_mon=12, tm_mday=21, tm_hour=13, tm_min=12, tm_sec=23, tm_wday=3, tm_yday=355, tm_isdst=-1)
```

Reply:
I think this is less precise, it doesn't check for zero padding and `%f` doesn't check for the max value.

Reply:
I think matching the zero-padding is overkill, but if you want to check for that, use a regex: 

```python
re.fullmatch(r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{9}", timestamp)
<re.Match object; span=(0, 29), match='2023-12-21T13:12:23.123456789'>
```

But for the rest, use `strptime` . It will do the right thing up to microseconds. If you want to also validate nanoseconds, you can just parse the remaining 3 characters: 

```python
assert int(timestamp[-3:]) >= 0
```